src/ng/compile.js

```diff
@@ -3101,6 +3130,27 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
         }
       });
 
+      function recordChanges(key, currentValue, previousValue) {
+        if (isFunction(destination.$onChanges)) {
+          // If we have not already scheduled the onChanges hook then do so now
+          if (!changes) {
+            changes = {};
+            scope.$$postDigest(triggerOnChangesHook);
+          }
+          // Store this change
+          changes[key] = {previousValue: previousValue, currentValue: currentValue};
+        }
+      }
+
+      function triggerOnChangesHook() {
+        // We must run this hook in an apply since the $$postDigest runs outside apply
+        scope.$apply(function() {
```

Comment:
I think we could move the login that calls `triggerOnChangeHook` inside `$compile` (I mean make it a compiler-local thing) and store the changes for **each** controller there, so we only trigger one extra `$digest`.